lifecycle-operator/webhooks/pod_mutator/handlers/pod_annotation.go

```diff
@@ -0,0 +1,125 @@
+package handlers
+
+import (
+	"context"
+
+	argov1alpha1 "github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/go-logr/logr"
+	apicommon "github.com/keptn/lifecycle-toolkit/lifecycle-operator/apis/lifecycle/v1alpha3/common"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type PodAnnotationHandler struct {
+	Client client.Client
+	Log    logr.Logger
+}
+
+func (p *PodAnnotationHandler) IsAnnotated(ctx context.Context, req admission.Request, pod *corev1.Pod) bool {
+	podIsAnnotated := isPodAnnotated(pod)
+	if !podIsAnnotated {
+		p.Log.Info("Pod is not annotated, check for parent annotations...")
+		podIsAnnotated = p.copyAnnotationsIfParentAnnotated(ctx, &req, pod)
+	}
+	return podIsAnnotated
+}
+
+func (p *PodAnnotationHandler) copyAnnotationsIfParentAnnotated(ctx context.Context, req *admission.Request, pod *corev1.Pod) bool {
+	podOwner := GetOwnerReference(&pod.ObjectMeta)
+	if podOwner.UID == "" {
+		return false
+	}
+
+	switch podOwner.Kind {
+	case "ReplicaSet":
+		rs := &appsv1.ReplicaSet{}
+		if err := p.Client.Get(ctx, types.NamespacedName{Namespace: req.Namespace, Name: podOwner.Name}, rs); err != nil {
+			return false
+		}
+		p.Log.Info("Done fetching RS")
+
+		rsOwner := GetOwnerReference(&rs.ObjectMeta)
+		if rsOwner.UID == "" {
+			return false
+		}
+
+		if rsOwner.Kind == "Rollout" {
+			ro := &argov1alpha1.Rollout{}
+			return p.fetchParentObjectAndCopyLabels(ctx, podOwner.Name, req.Namespace, pod, ro)
+		}
+		dp := &appsv1.Deployment{}
+		return p.fetchParentObjectAndCopyLabels(ctx, rsOwner.Name, req.Namespace, pod, dp)
+
+	case "StatefulSet":
+		sts := &appsv1.StatefulSet{}
+		return p.fetchParentObjectAndCopyLabels(ctx, podOwner.Name, req.Namespace, pod, sts)
+	case "DaemonSet":
+		ds := &appsv1.DaemonSet{}
+		return p.fetchParentObjectAndCopyLabels(ctx, podOwner.Name, req.Namespace, pod, ds)
+	default:
+		return false
+	}
+}
+
+func (p *PodAnnotationHandler) fetchParentObjectAndCopyLabels(ctx context.Context, name string, namespace string, pod *corev1.Pod, objectContainer client.Object) bool {
+	if err := p.Client.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, objectContainer); err != nil {
+		return false
+	}
+	objectContainerMetaData := metav1.ObjectMeta{
+		Labels:      objectContainer.GetLabels(),
+		Annotations: objectContainer.GetAnnotations(),
+	}
+	return copyResourceLabelsIfPresent(&objectContainerMetaData, pod)
```

Comment:
let's not call `copyResourceLabelsIfPresent` from inside of this function. it will be easier to test it if this function will return objectContainerMetaData and then 

`copyResourceLabelsIfPresent(&objectContainerMetaData, pod)` will be called

it will be easier to test is separately and it will make the function simpler and with a single thing to do